Add `--defaults` flag to allow loading of MySQL configuration

ChrisWiegman opened the pull request:
This fix addresses issue #155 

**The Problem**

The `--no-defaults` flag tells MySQL to ignore all .cnf files passed to it. While this works great in many situations, there are scenarios where it can break MySQL such as when the .cnf specifies an alternative socket file or other environment specific configuration items.

**The Proposed Solution**

This PR, after talking with @schlessera, fixes the issue by re-introducind the `--defaults` flag to the wp db command. When present, the _--no-defaults_ flag normally passed to all MySQL commands is removed thereby allowing MySQL to make use of the MYSQL_HOME or other config variables.

schlessera commented on README.md:
```diff
@@ -181,6 +181,9 @@ for more details on the `CHECK TABLE` statement.
 		Extra arguments to pass to mysqlcheck. [Refer to mysqlcheck docs](https://dev.mysql.com/doc/en/mysqlcheck.html).
 
+	[--defaults]
+		Removes the "--no-defaults" flag normally passed to MySQL allowing it to use the default my.cnf or one specified with the MYSQL_HOME environment variable.
```
The help text should focus less on the current implementation, and not use this "double negation" in the sense of "remove => no".

How about something like this:
```
Loads the environment's MySQL option files. Default behavior is to skip loading them to avoid failures due to misconfiguration.
```

ChrisWiegman replied:
Fixed

schlessera commented on features/db-check.feature:
```diff
@@ -97,3 +97,18 @@ Feature: Check the database
+    Then STDERR should contain:
+      """
+      Debug (db): /usr/bin/env mysqlcheck --no-defaults %s
+      """
```
Also add cases with the `--no-defaults` flag to make sure the extra logic still keeps the `--flag/--no-flag` behavior consistent.

So, `wp db check --no-defaults --debug` should have the same result right now as `wp db check --debug`.

ChrisWiegman replied:
Good call. I've added cases for each of the files in which I was also checking for the command.

schlessera replied:
I don't see these added yet, @ChrisWiegman - I was referring to actually using the flag `--no-defaults` in the command that you `When I try ...`, to ensure the flag handling is not broken with regards to negating it with `--no-...`.

schlessera commented on src/DB_Command.php:
```diff
@@ -198,15 +198,22 @@ public function clean( $_, $assoc_args ) {
 	public function check( $_, $assoc_args ) {
+
+		$command = sprintf( '/usr/bin/env mysqlcheck%s %s', self::get_no_defaults( $assoc_args ), '%s' );
+		WP_CLI::debug( $command, 'db' );
```
Add a qualifier to the raw `$command` variable here:
```suggestion
		WP_CLI::debug( "Running shell command: {$command}", 'db' );
```

schlessera commented on src/DB_Command.php:
```diff
@@ -1632,4 +1679,21 @@ private static function get_mysql_args( $assoc_args ) {
+	 * @param array $assoc_args Associative args array.
+	 * @return string Either the '--no-defaults' flag for use in the command or an empty string.
+	 */
+	private static function get_no_defaults( &$assoc_args ) {
```
I think this method name doesn't read well and might be confusing (i.e., we're not actually fetching or not fetching any "defaults").

I'd suggest using something like this instead:
```suggestion
	private static function get_defaults_flag_string( &$assoc_args ) {
```

schlessera replied:
Also, please use a regular protected method (not a private static one), and access it via `$this->`.

- I don't know why the helper methods in here are static, but that is not something to replicate, as static state is always a liability and should only be used if actually necessary. I'll fix the others in a subsequent PR.

- Using `private` within a class that is not `final` is a latent bug. Anyone extending this class and trying to override one of its public methods will not have access to the private methods it depends on. If they then try to override the private methods, they'll end up with two different versions of the method.

- Using `self` within a class that is not `final` is a latent bug as well. If someone overrides the class, any reference to `self` in the existing code will not take the extending class into account. `self` should be replaced with `static` in that case, to make use of late-static-binding.

schlessera commented on features/db-check.feature:
```diff
@@ -97,3 +97,18 @@ Feature: Check the database
       """
     And STDOUT should be empty
+
+  Scenario: Ensure MySQL defaults are available when as appropriate with --defaults flag
```
All of the tests are in one way or the other to "ensure" something, so that term is superfluous here.

```suggestion
  Scenario: MySQL defaults are available as appropriate with --defaults flag
```

ChrisWiegman commented:
@schlessera, I do believe I've addressed all your concerns. Per the static method issue, I had shared the concern but went with the existing paradigm. The new fixes change that but, note, also required a change to another private static method that had been calling the flags method. All tests still work, but I did want to call that out just in case. Please let me know if you find any further issues.

ChrisWiegman commented:
@schlessera, thank you for the clarification on the negation of flags. I didn't realize that from your initial comment. I've updated the flag login and added tests as appropriate.

schlessera commented:
@ChrisWiegman The PR works now and does what it should, thanks for keeping at it!

When opening it up in PHPStorm (and seeing more of the big picture of the file than here in the GitHub PR view), I realized that the file currently is a mess in terms of how it handles MySQL arguments, with one command passing all of them through, most others passing a subset, ...

As this is not directly related to your PR, I'll merge yours now, and then create a separate one to refactor a bit, so don't be surprised if I'm immediately swapping things around again after this was merged.

(Note to self: I should always open PRs in PHPStorm. It seems so convenient to only look in here, but you easily miss the bigger problems as large parts are just hidden by GH)

schlessera commented:
Thanks for the PR, @ChrisWiegman !
